airflow/providers/hashicorp/secrets/vault.py

```diff
@@ -0,0 +1,197 @@
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        role_id: Optional[str] = None,
+        secret_id: Optional[str] = None,
+        gcp_key_path: Optional[str] = None,
+        gcp_scopes: Optional[str] = None,
+        **kwargs
+    ):
+        super().__init__()
+        self.path = path.rstrip('/')
+        self.url = url
+        self.auth_type = auth_type
+        self.kwargs = kwargs
+        self.token = token
+        self.username = username
+        self.password = password
+        self.role_id = role_id
+        self.secret_id = secret_id
+        self.mount_point = mount_point
+        self.gcp_key_path = gcp_key_path
+        self.gcp_scopes = gcp_scopes
+        self._client: Optional[hvac.Client] = None
+
+    def get_client(self) -> hvac.Client:
+        """
+        Return an authenticated Hashicorp Vault client
+        """
+        if not self._client:
+            self._client = hvac.Client(url=self.url, **self.kwargs)
+            if self.auth_type == "token":
+                self._client.token = self.token
+            elif self.auth_type == "ldap":
+                self._client.auth.ldap.login(
+                    username=self.username, password=self.password)
+            elif self.auth_type == "userpass":
+                self._client.auth_userpass(username=self.username, password=self.password)
+            elif self.auth_type == "approle":
+                self._client.auth_approle(role_id=self.role_id, secret_id=self.secret_id)
+            elif self.auth_type == "github":
+                self._client.auth.github.login(token=self.token)
+            elif self.auth_type == "gcp":
+                credentials = self._get_gcp_credentials()
+                self._client.auth.gcp.configure(credentials=credentials)
+            else:
+                raise AirflowException(f"Authentication type '{self.auth_type}' not supported")
+
+            if self._client.is_authenticated():
+                return self._client
+            else:
+                raise VaultError("Vault Authentication Error!")
+        else:
+            return self._client
+
+    def get_conn_uri(self, conn_id: str) -> str:
+        """
+        Get secret value from Vault. Store the secret in the form of URI
+
+        :param conn_id: connection id
+        :type conn_id: str
+        """
+        connection_id = f"{CONN_ENV_PREFIX}{conn_id}".upper()
+        client = self.get_client()
+
+        self.log.debug("Path: %s", self.path)
+        self.log.debug("Mount Point: %s", self.mount_point)
+        self.log.debug("Retrieving the secret for Connection ID: %s", connection_id)
+
+        response = client.secrets.kv.v2.read_secret_version(
+            path=self.path, mount_point=self.mount_point)
+        return_data = response["data"]["data"]
+
+        try:
+            value = return_data[connection_id]
+            self.log.debug("Value of the secret: %s", return_data)
+        except KeyError:
```

Comment:
@kaxil if you raise here, then i think airflow will not continue to search in env vars then in metastore; it will just stop here and raise.

see https://github.com/apache/airflow/blob/master/airflow/secrets/__init__.py#L62

Reply:
Good point, will update

Reply:
Updated

Reply:
Also added a test case to capture that fallback on env vars